src/app/icd/IcdManagementServer.cpp

```diff
@@ -40,8 +41,17 @@ Status IcdManagementServer::RegisterClient(PersistentStorageDelegate & storage,
     // Save
     entry.checkInNodeID    = node_id;
     entry.monitoredSubject = monitored_subject;
-    entry.key              = key;
-    err                    = table.Set(entry.index, entry);
+    err = entry.SetKey(key);
+    VerifyOrReturnError(CHIP_ERROR_INVALID_ARGUMENT != err, InteractionModel::Status::ConstraintError);
+    VerifyOrReturnError(CHIP_NO_ERROR == err, InteractionModel::Status::Failure);
+    err = table.Set(entry.index, entry);
+
+    // Delete key upon failure to prevent key storage leakage.
+    if (err != CHIP_NO_ERROR)
+    {
+        entry.DeleteKey();
+    }
```

Comment:
Just to confirm, we delete the key only if setting the entry in the table fails?

Reply:
yes, otherwise the key Handle would be lost but the key would still exist in a secure storage..